File: Sources/Gamified/Services/GamifiedService.swift
//
//  GamifiedService.swift
//  
//
//  Created by Nathan Fallet on 28/01/2023.
//

import Foundation

public class GamifiedService {
    
    // MARK: - Storage
    
    private let globalStorage: GlobalStorageService
    private let dailyStorage: DailyStorageService
    
    /// Registered stats
    public private(set) var registeredStats = [RegisteredStats]()
    
    /// Registered achievements
    public private(set) var registeredAchievements = [RegisteredAchievement]()
    
    // MARK: - Initializer
    
    public init(
        globalStorage: GlobalStorageService,
        dailyStorage: DailyStorageService,
        registeredStats: [RegisteredStats],
        registeredAchievements: [RegisteredAchievement]
    ) {
        self.globalStorage = globalStorage
        self.dailyStorage = dailyStorage
        
        registeredStats.forEach(registerStats)
        registeredAchievements.forEach(registerAchievement)
    }
    
    // MARK: - Stats
    
    /// Get a stats
    /// - Parameters:
    ///   - key: The stats to get
    ///   - date: The day to fetch
    /// - Returns: The value of the stats
    public func getStats(_ key: String, for date: Date = Date()) -> Int {
        return dailyStorage.getValue(key, for: date)
    }
    
    /// Set a stats for a day
    /// - Parameters:
    ///   - key: The stats to set
    ///   - value: The value to set
    ///   - date: The day to set
    public func setStats(_ key: String, value: Int, for date: Date = Date()) {
        dailyStorage.setValue(key, value: value, for: date)
    }
    
    /// Increment a stats for a day
    /// - Parameters:
    ///   - key: The stats to increment
    ///   - value: The value to add
    ///   - date: The day to set
    public func incrementStats(_ key: String, by value: Int = 1, for date: Date = Date()) {
        setStats(key, value: getStats(key, for: date) + value, for: date)
    }
    
    /// Register a stats
    /// - Parameter key: The stats to register
    public func registerStats(_ key: RegisteredStats) {
        dailyStorage.setupValue(key.key)
        registeredStats.append(key)
    }
    
    /// Get a graph for stats
    /// - Parameters:
    ///   - key: The stats to get
    ///   - startDate: The starting date
    ///   - endDate: The ending date
    /// - Returns: The graph corresponding to this stats between startDate and endDate
    public func getGraph(_ key: String, startDate: Date, endDate: Date) -> Graph {
        return Graph(
            key: key,
            title: registeredStats.first(where: { $0.key == key })?.name ?? "Unknown",
            stats: dailyStorage
                .getValues(key, startDate: startDate, endDate: endDate)
                .map { date, value in
                    Stats(day: date, value: value)
                }
        )
    }
    
    /// Get graphs' stats
    /// - Parameters:
    ///   - startDate: The starting date
    ///   - endDate: The ending date
    /// - Returns: All graphs' stats between startDate and endDate
    public func getGraphs(startDate: Date, endDate: Date) -> [Graph] {
        return registeredStats.map { key in
            getGraph(key.key, startDate: startDate, endDate: endDate)
        }
    }
    
    // MARK: - Values
    
    /// Get a value
    /// - Parameter key: The value to get
    /// - Returns: The value
    public func getValue(_ key: String) -> Int {
        return globalStorage.getValue(key)
    }
    
    /// Set a value
    /// - Parameters:
    ///   - key: The value to set
    ///   - value: The value to set
    public func setValue(_ key: String, value: Int) {
        globalStorage.setValue(key, value: value)
    }
    
    /// Increment a value
    /// - Parameters:
    ///   - key: The value to increment
    ///   - value: The value to add
    public func incrementValue(_ key: String, by value: Int = 1) {
        setValue(key, value: getValue(key) + value)
    }
    
    // MARK: - Experience
    
    /// Get information about experience and level
    /// - Returns: An object with all that information
    public func getExperience() -> Experience {
        return Experience(total: getValue("internal_experience"))
    }
    
    @discardableResult
    /// Add experience to current level
    /// - Parameter exp: The amount of experience to add
    /// - Returns: A tuple containing experience information before and after that experience is added
    public func gainExperience(_ exp: Int) -> (Experience, Experience) {
        let previousExperience = getExperience()
        let newExperience = Experience(total: previousExperience.total + exp)
        setValue("internal_experience", value: newExperience.total)
        NotificationCenter.default.post(
            name: .experienceGained,
            object: nil,
            userInfo: [
                "previousExperience": previousExperience,
                "newExperience": newExperience
            ]
        )
        
        return (previousExperience, newExperience)
    }
    
    // MARK: - Achievements
    
    /// Register an achievement
    /// - Parameter key: The achievement to register
    public func registerAchievement(_ key: RegisteredAchievement) {
        registeredAchievements.append(key)
    }
    
    /// Get actual data for registered achievements
    /// - Returns: Data for registered achievements
    public func getAchievements() -> [Achievement] {
        return registeredAchievements.map { key in
            Achievement(
                key: key.key,
                icon: key.icon,
                text: key.name,
                value: getValue(key.key),
                target: key.target
            )
        }
    }
    
}

File: Gamified/src/main/java/me/nathanfallet/gamified/services/GamifiedService.kt
package me.nathanfallet.gamified.services

import me.nathanfallet.gamified.events.AchievementUnlockedEvent
import me.nathanfallet.gamified.events.ExperienceGainedEvent
import me.nathanfallet.gamified.models.Achievement
import me.nathanfallet.gamified.models.Experience
import me.nathanfallet.gamified.models.Graph
import me.nathanfallet.gamified.models.RegisteredAchievement
import me.nathanfallet.gamified.models.RegisteredStats
import me.nathanfallet.gamified.models.Stats
import org.greenrobot.eventbus.EventBus
import java.util.Date

class GamifiedService(
    private val globalStorage: GlobalStorageService,
    private val dailyStorage: DailyStorageService,
    registeredStats: List<RegisteredStats>,
    registeredAchievements: List<RegisteredAchievement>
) {

    // MARK: - Shared instance

    companion object {

        /**
         * Shared instance of the service, used by view models
         */
        @JvmStatic
        lateinit var shared: GamifiedService

    }

    // MARK: - Storage

    private val _registeredStats = mutableListOf<RegisteredStats>()
    private val _registeredAchievements = mutableListOf<RegisteredAchievement>()

    /**
     * Registered stats
     */
    val registeredStats: List<RegisteredStats>
        get() = _registeredStats

    /**
     * Registered achievements
     */
    val registeredAchievements: List<RegisteredAchievement>
        get() = _registeredAchievements

    // MARK: - Initializer

    init {
        registeredStats.forEach(this::registerStats)
        registeredAchievements.forEach(this::registerAchievement)
    }

    // MARK: - Stats

    /**
     * Get a stats for a day
     * @param key The stats to get
     * @param date The day to get
     * @return The value of this stats for this day
     */
    fun getStats(key: String, date: Date = Date()): Long {
        return dailyStorage.getValue(key, date)
    }

    /**
     * Set a stats for a day
     * @param key The stats to set
     * @param value The value to set
     * @param date The day to set
     */
    fun setStats(key: String, value: Long, date: Date = Date()) {
        dailyStorage.setValue(key, value, date)
    }

    /**
     * Increment a stats for a day
     * @param key The stats to increment
     * @param value The value to increment
     * @param date The day to increment
     */
    fun incrementStats(key: String, value: Long = 1, date: Date = Date()) {
        setStats(key, getStats(key, date) + value, date)
    }

    /**
     * Register a stats
     * @param key The stats to register
     */
    fun registerStats(key: RegisteredStats) {
        dailyStorage.setupValue(key.key)
        _registeredStats.add(key)
    }

    /**
     * Get a graph for stats
     * @param key: The stats to get
     * @param startDate: The starting date
     * @param endDate: The ending date
     * @return The graph corresponding to this stats between startDate and endDate
     */
    fun getGraph(key: String, startDate: Date, endDate: Date): Graph {
        return Graph(
            key,
            registeredStats.firstOrNull { it.key == key }?.name ?: "Unknown",
            dailyStorage.getValues(key, startDate, endDate).map { Stats(it.key, it.value) }
        )
    }

    /**
     * Get all graphs' stats
     * @param startDate: The starting date
     * @param endDate: The ending date
     * @return All graphs' stats between startDate and endDate
      */
    fun getGraphs(startDate: Date, endDate: Date): List<Graph> {
        return registeredStats.map { getGraph(it.key, startDate, endDate) }
    }

    // MARK: - Values

    /**
     * Get a value
     * @param key The value to get
     * @return The value
     */
    fun getValue(key: String): Long {
        return globalStorage.getValue(key)
    }

    /**
     * Set a value
     * @param key The value to set
     * @param value The value to set
     * @param incrementing: If the value should be added or replaced
     */
    fun setValue(key: String, value: Long, incrementing: Boolean = false) {
        // Save value
        val previousValue = getValue(key)
        val newValue = if (incrementing) previousValue + value else value
        globalStorage.setValue(key, newValue)

        // Check for achievements
        registeredAchievements.filter {
            key == it.key && previousValue < it.target && newValue >= it.target
        }.forEach { achievement ->
            EventBus.getDefault().post(AchievementUnlockedEvent(achievement))
            gainExperience(achievement.experience)
        }
    }

    /**
     * Increment a value
     * @param key The value to increment
     * @param value The value to add
     */
    fun incrementValue(key: String, value: Long = 1) {
        setValue(key, value, true)
    }

    // MARK: - Experience

    /**
     * Get information about experience and level
     * @return An object with all that information
     */
    fun getExperience(): Experience {
        return Experience(getValue("internal_experience"))
    }

    /**
     * Add experience to current level
     * @param exp The amount of experience to add
     * @return A pair containing experience information before and after that experience is added
     */
    fun gainExperience(exp: Long): Pair<Experience, Experience> {
        val previousExperience = getExperience()
        val newExperience = Experience(previousExperience.total + exp)
        setValue("internal_experience", newExperience.total)
        EventBus.getDefault().post(ExperienceGainedEvent(previousExperience, newExperience))

        return Pair(previousExperience, newExperience)
    }

    // MARK: - Achievements

    /**
     * Register an achievement
     * @param key: The achievement to register
     */
    fun registerAchievement(key: RegisteredAchievement) {
        _registeredAchievements.add(key)
    }

    /**
     * Get actual data for registered achievements
     * @return Data for registered achievements
     */
    fun getAchievements(): List<Achievement> {
        return registeredAchievements.map {
            Achievement(
                it.key,
                it.icon,
                it.name,
                getValue(it.key),
                it.target
            )
        }
    }

}